# Waveform generalization

Evaluates how trained policies behave under alternate force waveforms and input signals.
The notebook reads model/evaluation artifacts produced by the shared experiment entry points.


# Waveform Generalization

This notebook renders waveform generalization results inline for both the baseline evaluation and the warm-start curriculum.

In [1]:
import sys
from pathlib import Path

for candidate in [Path.cwd().resolve(), *Path.cwd().resolve().parents]:
    if (candidate / '.git').exists() and (candidate / 'matlab_env_python_replica').exists():
        if str(candidate) not in sys.path:
            sys.path.insert(0, str(candidate))
        break

from notebooks._teleop_nb import repo_paths, show_rows

P = repo_paths()
DQN_DYN = P['dqn_results'] / 'dyn'
show_rows([
    {'run': run_name, 'exists': (DQN_DYN / run_name).exists(), 'path': str(DQN_DYN / run_name)}
    for run_name in ['dqn_legacy_baseline_fast20s_5N_01', 'dqn_waveform_curriculum_warmstart_5N_01']
], title='Generalization runs', max_rows=10)


### Generalization runs

run,exists,path
dqn_legacy_baseline_fast20s_5N_01,True,C:\Users\aha173\OneDrive - American University of Beirut\AGV research\spring VIPP obsidian notes\VIPP\VIPP naseem\VIPP\Teleop\TeleopWithRL\matlab_env_python_replica\dqn_experiments\results\dyn\dqn_legacy_baseline_fast20s_5N_01
dqn_waveform_curriculum_warmstart_5N_01,True,C:\Users\aha173\OneDrive - American University of Beirut\AGV research\spring VIPP obsidian notes\VIPP\VIPP naseem\VIPP\Teleop\TeleopWithRL\matlab_env_python_replica\dqn_experiments\results\dyn\dqn_waveform_curriculum_warmstart_5N_01


In [2]:
from notebooks._teleop_nb import show_csv

baseline_waveform_root = DQN_DYN / 'dqn_legacy_baseline_fast20s_5N_01' / '00b' / 'dqn' / 'wv'
curriculum_waveform_root = DQN_DYN / 'dqn_waveform_curriculum_warmstart_5N_01' / '1w'

show_csv(baseline_waveform_root / 'waveform_summary.csv', title='Baseline waveform summary', max_rows=20)
show_csv(curriculum_waveform_root / 'waveform_summary.csv', title='Curriculum waveform summary', max_rows=20)


### Baseline waveform summary

### `C:\Users\aha173\OneDrive - American University of Beirut\AGV research\spring VIPP obsidian notes\VIPP\VIPP naseem\VIPP\Teleop\TeleopWithRL\matlab_env_python_replica\dqn_experiments\results\dyn\dqn_legacy_baseline_fast20s_5N_01\00b\dqn\wv\waveform_summary.csv`

waveform,waveform_dir,model_path,label,test_episodes,force_amp,force_bias,force_freq_rad,force_phase,mean_reward,tracking_rmse_m,transparency_rmse_w,pre_switch_tracking_rmse_m,post_switch_tracking_rmse_m,pre_switch_transparency_rmse_w,post_switch_transparency_rmse_w,invalid_episode
sine,sin,TeleopWithRL\matlab_env_python_replica\dqn_experiments\results\dyn\dqn_legacy_baseline_fast20s_5N_01\00b\dqn\m\dqn_model.pt,DQN_baseline_S0_baseline_full10_baseline_cfg,100,5.0,5.0,0.5,0.0,-68.13277949021573,0.0098805281027806,3.271772957293813,0.009691061312439832,0.0037366909970588114,3.327354961610141,0.5904410410901462,0.08
cosine,cos,TeleopWithRL\matlab_env_python_replica\dqn_experiments\results\dyn\dqn_legacy_baseline_fast20s_5N_01\00b\dqn\m\dqn_model.pt,DQN_baseline_S0_baseline_full10_baseline_cfg,100,5.0,5.0,0.5,0.0,-140.15233050741963,0.00723916243026021,0.9034348111445907,0.00706380947118745,0.0073394820605135426,0.9866722506455373,0.7471668814352377,0.0
square,sqr,TeleopWithRL\matlab_env_python_replica\dqn_experiments\results\dyn\dqn_legacy_baseline_fast20s_5N_01\00b\dqn\m\dqn_model.pt,DQN_baseline_S0_baseline_full10_baseline_cfg,100,5.0,5.0,0.5,0.0,-214.07602373593744,0.011185800357392594,2.393600602304093,0.011185800357392594,0.0,2.393600602304093,0.0,1.0
ramp,rmp,TeleopWithRL\matlab_env_python_replica\dqn_experiments\results\dyn\dqn_legacy_baseline_fast20s_5N_01\00b\dqn\m\dqn_model.pt,DQN_baseline_S0_baseline_full10_baseline_cfg,100,5.0,5.0,0.5,0.0,-263.2854815003319,0.01541515672544376,3.597891882984243,0.008488207997956127,0.01327052609577085,2.981158992127399,1.6106774523672172,0.08
multisine,mul,TeleopWithRL\matlab_env_python_replica\dqn_experiments\results\dyn\dqn_legacy_baseline_fast20s_5N_01\00b\dqn\m\dqn_model.pt,DQN_baseline_S0_baseline_full10_baseline_cfg,100,5.0,5.0,0.5,0.0,-1462807.105420303,0.008354705234747626,1082.4779521070798,0.008142443249274114,0.0032708130065235356,1082.4928386195336,0.702643877128267,0.08


### Curriculum waveform summary

### `C:\Users\aha173\OneDrive - American University of Beirut\AGV research\spring VIPP obsidian notes\VIPP\VIPP naseem\VIPP\Teleop\TeleopWithRL\matlab_env_python_replica\dqn_experiments\results\dyn\dqn_waveform_curriculum_warmstart_5N_01\1w\waveform_summary.csv`

waveform,waveform_dir,model_path,label,train_waveforms,test_episodes,force_amp,force_bias,force_freq_rad,force_phase,mean_reward,tracking_rmse_m,transparency_rmse_w,pre_switch_tracking_rmse_m,post_switch_tracking_rmse_m,pre_switch_transparency_rmse_w,post_switch_transparency_rmse_w,invalid_episode
sine,sin,C:\Users\aha173\OneDrive - American University of Beirut\AGV research\spring VIPP obsidian notes\VIPP\VIPP naseem\VIPP\Teleop\TeleopWithRL\matlab_env_python_replica\dqn_experiments\results\dyn\dqn_waveform_curriculum_warmstart_5N_01\0t\dqn\m\dqn_model.pt,DQN_curriculum_S0_baseline_full10_baseline_cfg,"['sine', 'cosine', 'square', 'ramp', 'multisine']",100,5.0,5.0,0.5,0.0,-41.64579449450305,0.0036256672611859964,0.7171967363553362,0.003545895913882635,0.00365717526686709,0.855983630113615,0.5114522030340596,0.0
cosine,cos,C:\Users\aha173\OneDrive - American University of Beirut\AGV research\spring VIPP obsidian notes\VIPP\VIPP naseem\VIPP\Teleop\TeleopWithRL\matlab_env_python_replica\dqn_experiments\results\dyn\dqn_waveform_curriculum_warmstart_5N_01\0t\dqn\m\dqn_model.pt,DQN_curriculum_S0_baseline_full10_baseline_cfg,"['sine', 'cosine', 'square', 'ramp', 'multisine']",100,5.0,5.0,0.5,0.0,-44.42120853688373,0.003745976943486686,0.6973836964501239,0.0035421844879589327,0.003876792457031578,0.81957551971898,0.5022585943379151,0.0
square,sqr,C:\Users\aha173\OneDrive - American University of Beirut\AGV research\spring VIPP obsidian notes\VIPP\VIPP naseem\VIPP\Teleop\TeleopWithRL\matlab_env_python_replica\dqn_experiments\results\dyn\dqn_waveform_curriculum_warmstart_5N_01\0t\dqn\m\dqn_model.pt,DQN_curriculum_S0_baseline_full10_baseline_cfg,"['sine', 'cosine', 'square', 'ramp', 'multisine']",100,5.0,5.0,0.5,0.0,-38.282907811644066,0.002746158107183874,0.9414320397155005,0.002556048271506555,0.0028185110303771627,0.9340700767630474,0.8922825632250987,0.0
ramp,rmp,C:\Users\aha173\OneDrive - American University of Beirut\AGV research\spring VIPP obsidian notes\VIPP\VIPP naseem\VIPP\Teleop\TeleopWithRL\matlab_env_python_replica\dqn_experiments\results\dyn\dqn_waveform_curriculum_warmstart_5N_01\0t\dqn\m\dqn_model.pt,DQN_curriculum_S0_baseline_full10_baseline_cfg,"['sine', 'cosine', 'square', 'ramp', 'multisine']",100,5.0,5.0,0.5,0.0,-26.94221077557251,0.0024927218146733846,0.6408383087911274,0.0023993961426458836,0.00253516945276119,0.6032314019152966,0.625677917025325,0.0
multisine,mul,C:\Users\aha173\OneDrive - American University of Beirut\AGV research\spring VIPP obsidian notes\VIPP\VIPP naseem\VIPP\Teleop\TeleopWithRL\matlab_env_python_replica\dqn_experiments\results\dyn\dqn_waveform_curriculum_warmstart_5N_01\0t\dqn\m\dqn_model.pt,DQN_curriculum_S0_baseline_full10_baseline_cfg,"['sine', 'cosine', 'square', 'ramp', 'multisine']",100,5.0,5.0,0.5,0.0,-40.31778988171505,0.0035063362479597015,0.7120908309837103,0.003644819889170388,0.003330992575979255,0.8276568990890264,0.546536299977741,0.0


## Selected result

The warm-start curriculum DQN model is retained because it recorded zero invalid episodes across all five tested waveforms and the strongest aggregate errors in the executed notebook output. This is archival evidence from the embedded output because the original DQN result directory is not tracked in the current checkout.

| Waveform | Tracking RMSE [mm] | Transparency RMSE [W] | Pre-switch track [mm] | Post-switch track [mm] | Pre-switch transp. [W] | Post-switch transp. [W] | Invalid episodes |
|---|---:|---:|---:|---:|---:|---:|---:|
| sine | 3.626 | 0.717 | 3.546 | 3.657 | 0.856 | 0.511 | 0.0% |
| cosine | 3.746 | 0.697 | 3.542 | 3.877 | 0.820 | 0.502 | 0.0% |
| square | 2.746 | 0.941 | 2.556 | 2.819 | 0.934 | 0.892 | 0.0% |
| ramp | **2.493** | **0.641** | 2.399 | 2.535 | 0.603 | 0.626 | 0.0% |
| multisine | 3.506 | 0.712 | 3.645 | 3.331 | 0.828 | 0.547 | 0.0% |

![Waveform curriculum evaluation bars](../../results_index/figures/waveform_curriculum_evaluation_bars.png)
